api/lib/round-tracker.js

```diff
@@ -245,10 +252,30 @@ export async function maybeCreateSparkRound (pgClient, {
     meridianContractAddress,
     meridianRoundIndex,
     roundStartEpoch,
-    MAX_TASKS_PER_NODE
+    BASELINE_MAX_TASKS_PER_NODE
   ])
 
   if (rowCount) {
+    if (useDynamicTaskCount) {
+      try {
+        const res = await fetch('https://stats.filspark.com/stations/daily')
+        const [{ station_id_count: stationIdCount }] = await res.json()
```

Comment:
I don't think the number of stations per day is trustworthy data right now. We know there are three participants that are churning their Station IDs, adding 900k unique station IDs per day, where most of their station IDs contribute less than 10 measurements only.

> Base task count on measurement count from previous round. This data doesn't exist in spark-stats yet. It should lead to better results though, than the currently used station-id count. @bajtos wdyt? Shall we give this a try and see if it behaves well enough?

Yes, I agree it will be better to use the number of measurements instead of the number of stations.

But then, how do we convert the "desired number of measurements" to "tasks per node" when we don't know how many nodes will submit measurements in this round? We will never know that number, we have to estimate it based on the historical data.

I think it will be great to have a stateful algorithm that's watching how the number of measurements per round evolves from round to round.

Let's say we want to perform MD measurements _per day_. There are 24h/day * 60 minutes/hour / 20 minutes/round = 72 rounds per day. That means we want to record MD/72 measurements per round _averaged over the entire day_.

The logic for deciding on the number of measurements per node for the current round could work like this:

1. Check how many measurements were collected today = T
2. Calculate how many rounds are remaining today = 72 - roundsCompletedToday = RR
3. Now we can calculate the desired "tasks per node" as `(MD - MT) / RR`

The downside is that if the network misbehaves and records way too many measurements early during the day, we will have to perform significantly fewer measurements in the evening. We could possibly stop the network completely if it already exhausted the quote of how many measurements are allowed per day.

We are already storing stats about the number of measurements collected.

- https://stats.filspark.com/retrieval-success-rate returns the total number of all & accepted measurements
- https://stats.filspark.com/miners/retrieval-success-rate/summary returns all & accepted measurement counts with per-miner granularity.

The responses with today's data are cached by Cloudflare for 10 minutes - you may want to disable the cache in your requests.

Reply:
Yes, I was also thinking about this stateful algorithm, similarly to how you described. Only, I would not operate based on jobs per day, but jobs per round. The desired job count per round is X, depending on whether round n-1 over- or undershot the goal X, adjust the multiplier / offset from round n-1.

Reply:
> Yes, I was also thinking about this stateful algorithm, similarly to how you described. Only, I would not operate based on jobs per day, but jobs per round. The desired job count per round is X, depending on whether round n-1 over- or undershot the goal X, adjust the multiplier / offset from round n-1.

I think that's a good start, your proposal should work well if the number of active Stations is mostly the same every round. Which should be true since most of our network consists of server nodes running 24/7.

However, if there were twice as many nodes running in the UTC evening (America working hours) than in the UTC morning (Asia working hours), then your algorithm will produce more or fewer jobs than we want, depending on whether you calibrate X using the data from UTC morning or evening.

Reply:
If we calibrate it based on the previous round, then - while we will be 20m behind during a growth/decline phase - we won't be catastrophically off

Reply:
SGTM 👍🏻

Reply:
Question:

Currently, `spark-stats` operates on `?from` and `&to` filters that are dates or date-time strings.

At first I thought about making the new API `GET /measurements/round?from=roundIndex&to=roundIndex`, but realized I'm fighting against the system constantly.

Therefore I'm wondering whether it's better to have `GET /measurements/per-round?from=dateTime&to=dateTime`, and we track internally at which timestamp a round has started and ended. It's a bit awkward for the query logic (which will live in `spark-api`), as they know the round index, but round start / end times are not easy to synchronize.

From the consumer (`spark-api`) perspective, it would be easiest if the new API supported round index as query filters.

Wdyt @bajtos?

Reply:
Do you need to get values for more than one round? In https://github.com/filecoin-station/spark-api/pull/385#discussion_r1731110859, you proposed to use data from the previous round only. In that case, I think a new endpoint not using any query strings would be the best.

E.g.

```
GET /measurement-count/round/123
```

If you want to make this future proof and support a range of rounds, then yes, let's use round index in the query filters. I propose to use different param names to avoid confusion.

Note that we have a round index that's local to each smart contract version and a round number that's global to the entire Spark deployment.

Based on my past experience, I recommend using the pair (`smartContractAddress, smartContractRoundIndex`) to identify the round you are interested in. That's how "get current round details" works in spark-api.

E.g.

```
GET /measurement-counts?meridianVersion=0x8460766edc62b525fc1fa4d628fc79229dc73031&fromRoundIndex=14800&toRoundIndex=14817
```

Also note that it's super easy to bypass the code handling from/to. Just don't wrap your request handler with `createRespondWithFetchFn` and implement query string parsing and cache TTL yourself.

Reply:
Agreed, the question above is outdated wrt the comment linked to by you.

What do you think about removing the round index from the query? This way the consumer can be simpler, and it's impossible to create a wrong request. Just:

```
GET /last-round-measurement-count/
```

(Exact path name TBD). This endpoint doesn't need heavy caching, as - for now - we only know of spark-api wanting to use it, every 20 minutes.

> Note that we have a round index that's local to each smart contract version and a round number that's global to the entire Spark deployment.

Removing the round index also resolves this question.

> Also note that it's super easy to bypass the code handling from/to. Just don't wrap your request handler with createRespondWithFetchFn and implement query string parsing and cache TTL yourself.

Very useful point 👍